File: src/musikdroid/app/src/main/java/io/casey/musikcube/remote/playback/PlayerWrapper.java
package io.casey.musikcube.remote.playback;

import android.util.Log;

import java.util.HashSet;
import java.util.Set;

import io.casey.musikcube.remote.util.Preconditions;

public abstract class PlayerWrapper {
    private static final String TAG = "MediaPlayerWrapper";
    private static final float DUCK_COEF = 0.2f; /* volume = 20% when ducked */
    private static final float DUCK_NONE = -1.0f;

    public enum State {
        Stopped,
        Preparing,
        Prepared,
        Playing,
        Paused,
        Error,
        Finished,
        Killing,
        Disposed
    }

    public interface OnStateChangedListener {
        void onStateChanged(PlayerWrapper mpw, State state);
    }

    private static Set<PlayerWrapper> activePlayers = new HashSet<>();
    private static float globalVolume = 1.0f;
    private static boolean globalMuted = false;
    private static float preDuckGlobalVolume = DUCK_NONE;

    public static void duck() {
        Preconditions.throwIfNotOnMainThread();

        if (preDuckGlobalVolume == DUCK_NONE) {
            final float lastVolume = globalVolume;
            setGlobalVolume(globalVolume * DUCK_COEF);
            preDuckGlobalVolume = lastVolume;
        }
    }

    public static void unduck() {
        Preconditions.throwIfNotOnMainThread();

        if (preDuckGlobalVolume != DUCK_NONE) {
            final float temp = preDuckGlobalVolume;
            preDuckGlobalVolume = DUCK_NONE;
            setGlobalVolume(temp);
        }
    }

    public static void setGlobalVolume(float volume) {
        Preconditions.throwIfNotOnMainThread();

        if (preDuckGlobalVolume != DUCK_NONE) {
            preDuckGlobalVolume = volume;
            volume = volume * DUCK_COEF;
        }

        if (volume != globalVolume) {
            globalVolume = volume;
            for (final PlayerWrapper w : activePlayers) {
                w.updateVolume();
            }
        }
    }

    public static float getGlobalVolume() {
        Preconditions.throwIfNotOnMainThread();

        if (globalMuted) {
            return 0;
        }

        return (preDuckGlobalVolume == DUCK_NONE) ? globalVolume : preDuckGlobalVolume;
    }

    public static void setGlobalMute(final boolean muted) {
        Preconditions.throwIfNotOnMainThread();

        if (PlayerWrapper.globalMuted != muted) {
            PlayerWrapper.globalMuted = muted;

            for (final PlayerWrapper w : activePlayers) {
                w.updateVolume();
            }
        }
    }

    public static PlayerWrapper newInstance() {
        //return new MediaPlayerWrapper();
        return new ExoPlayerWrapper();
    }

    protected static void addActivePlayer(final PlayerWrapper player) {
        Preconditions.throwIfNotOnMainThread();
        activePlayers.add(player);
    }

    protected static void removeActivePlayer(final PlayerWrapper player) {
        Preconditions.throwIfNotOnMainThread();
        activePlayers.remove(player);
    }


    private OnStateChangedListener listener;
    private State state = State.Stopped;

    public abstract void play(final String uri);
    public abstract void prefetch(final String uri);
    public abstract void pause();
    public abstract void resume();
    public abstract void setPosition(int millis);
    public abstract int getPosition();
    public abstract int getDuration();
    public abstract void updateVolume();
    public abstract void setNextMediaPlayer(final PlayerWrapper wrapper);
    public abstract void dispose();

    public void setOnStateChangedListener(OnStateChangedListener listener) {
        Preconditions.throwIfNotOnMainThread();

        this.listener = listener;

        if (listener != null) {
            this.listener.onStateChanged(this, state);
        }
    }

    public final State getState() {
        return this.state;
    }

    protected void setState(final PlayerWrapper.State state) {
        if (this.state != state) {
            this.state = state;
            if (listener != null) {
                this.listener.onStateChanged(this, state);
            }
        }
    }
}

File: src/musikdroid/app/src/main/java/io/casey/musikcube/remote/playback/PlayerWrapper.kt
package io.casey.musikcube.remote.playback

import io.casey.musikcube.remote.Application
import io.casey.musikcube.remote.data.ITrack
import io.casey.musikcube.remote.db.offline.OfflineTrack
import io.casey.musikcube.remote.util.Preconditions
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import org.json.JSONObject
import java.util.*

abstract class PlayerWrapper {
    private enum class Type {
        MediaPlayer, ExoPlayer
    }

    enum class State {
        Stopped,
        Preparing,
        Prepared,
        Playing,
        Buffering,
        Paused,
        Error,
        Finished,
        Killing,
        Disposed
    }

    private var listener: ((PlayerWrapper, State) -> Unit)? = null

    var state = State.Stopped
        protected set(state) {
            if (this.state != state) {
                field = state
                listener?.invoke(this,state)
            }
        }

    abstract fun play(uri: String, metadata: ITrack)
    abstract fun prefetch(uri: String, metadata: ITrack)
    abstract fun pause()
    abstract fun resume()
    abstract fun updateVolume()
    abstract fun setNextMediaPlayer(wrapper: PlayerWrapper?)
    abstract fun dispose()
    abstract var position: Int
    abstract val duration: Int
    abstract val bufferedPercent: Int

    open fun setOnStateChangedListener(listener: ((PlayerWrapper, State) -> Unit)?) {
        Preconditions.throwIfNotOnMainThread()
        this.listener = listener
        this.listener?.invoke(this, this.state)
    }

    companion object {
        private val TYPE = Type.ExoPlayer
        private val DUCK_COEF = 0.2f /* volume = 20% when ducked */
        private val DUCK_NONE = -1.0f

        private val activePlayers = HashSet<PlayerWrapper>()
        private var globalVolume = 1.0f
        private var globalMuted = false
        private var preDuckGlobalVolume = DUCK_NONE

        fun storeOffline(uri: String, metadata: ITrack) {
            Single.fromCallable {
                val track = OfflineTrack()
                if (track.fromJSONObject(uri, metadata.toJson())) {
                    Application.offlineDb?.trackDao()?.insertTrack(track)
                    Application.offlineDb?.prune()
                }
            }
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe()
        }

        fun duck() {
            Preconditions.throwIfNotOnMainThread()

            if (preDuckGlobalVolume == DUCK_NONE) {
                val lastVolume = globalVolume
                setVolume(globalVolume * DUCK_COEF)
                preDuckGlobalVolume = lastVolume
            }
        }

        fun unduck() {
            Preconditions.throwIfNotOnMainThread()

            if (preDuckGlobalVolume != DUCK_NONE) {
                val temp = preDuckGlobalVolume
                preDuckGlobalVolume = DUCK_NONE
                setVolume(temp)
            }
        }

        fun setVolume(volume: Float) {
            var computedVolume = volume
            Preconditions.throwIfNotOnMainThread()

            if (preDuckGlobalVolume != DUCK_NONE) {
                preDuckGlobalVolume = volume
                computedVolume *= DUCK_COEF
            }

            if (computedVolume != globalVolume) {
                globalVolume = computedVolume
                for (w in activePlayers) {
                    w.updateVolume()
                }
            }
        }

        fun getVolume(): Float {
            Preconditions.throwIfNotOnMainThread()

            if (globalMuted) {
                return 0f
            }

            return if (preDuckGlobalVolume == DUCK_NONE) globalVolume else preDuckGlobalVolume
        }

        fun setMute(muted: Boolean) {
            Preconditions.throwIfNotOnMainThread()

            if (PlayerWrapper.globalMuted != muted) {
                PlayerWrapper.globalMuted = muted

                for (w in activePlayers) {
                    w.updateVolume()
                }
            }
        }

        fun newInstance(): PlayerWrapper {
            return if (TYPE == Type.ExoPlayer)
                ExoPlayerWrapper() else MediaPlayerWrapper()
        }

        fun addActivePlayer(player: PlayerWrapper) {
            Preconditions.throwIfNotOnMainThread()
            activePlayers.add(player)
        }

        fun removeActivePlayer(player: PlayerWrapper) {
            Preconditions.throwIfNotOnMainThread()
            activePlayers.remove(player)
        }
    }
}
